doc/development/powershell.md

```diff
@@ -0,0 +1,165 @@
+This page contains guidelines for developing or updating powershell scripts used by the Azure SDK engineering system.
+
+Table of Contents
+=================
+
+* [TLDR](#tldr)
+* [Structure](#structure)
+* [Functionality](#functionality)
+* [Style](#style)
+* [Testing](#testing)
+   * [Unit Testing](#unit-testing)
+      * [Running Pester Tests](#running-pester-tests)
+   * [Local/Pipeline Functional Testing](#localpipeline-functional-testing)
+
+## TLDR
+
+_(i.e. too long, didn't read)_
+
+**DO**
+
+- Try to write scripts that can be run locally
+- Write unit tests for your scripts
+- Handle exit codes from external programs
+- Use `Set-StrictMode -Version 4` where possible
+- Define your data structures as much as possible via composition, or unit test or comment examples.
+- Use an editor extension or `Invoke-ScriptAnalyzer` to analyze code.
+
+**DON'T**
+
+- Write medium-large scripts inline to azure pipelines yaml (unless absolutely necessary)
+- Write code without putting logic into functions. Scripts grow in size and complexity over time, and if they start out un-organized, they tend to stay un-organized.
+
+## Structure
+
+When structuring a powershell cmdlet or collection of scripts, it is important to think about how easy that code will be to re-use and to test (primarily unit tests and/or running locally). 
+
+- When writing a powershell cmdlet file, keep the file as small as possible, or organize your code into functions. Scripts tend to grow in size and complexity over time with incremental edits from many authors, and it is much easier to understand, test and refactor code when it is written in a modular way from the start.
+- Isolate calls to dependencies inside their own cmdlet files or within functions. This makes it easier to abstract these calls for testing.
+- Avoid global state where possible. Powershell has very liberal scoping rules, and it is easy to take dependencies on variables that are defined outside a code block, which can lead to hard to debug issues. Where possible, only treat script parameters and variables defined at the top of the file as read-only global variables.
+- Try not to rely on exporting your code as a powershell module. Favor [dot sourcing](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scripts?view=powershell-7.2#script-scope-and-dot-sourcing) when possible. There are lots of downsides to working with module imports in local development settings that make them not worth it. This guideline some limitations, e.g. if you want to test class instance assertions via Pester, you cannot export the class types without using module imports.
+
+## Functionality
```

Comment:
Developers should also not rely on aliases. For example, `foreach` is also an alias for `ForEach-Object`. Depending on how you use it, it will be one or the other.

While any cmdlet can be overridden and impossible to reference explicitly (unless part of a module; see my comment above), it's far more common that user environments would override aliases.

This is actually best practice guidance PowerShell has given in the past.

Reply:
By `foreach` I mean https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_foreach?view=powershell-7.2#syntax. 

For other aliases, there is a bullet in `Style` around not using aliases (echo, %).

Reply:
I was referencing `foreach` only as an example. In fact, below you also say not to use `?` (`where-object`) or `%` (`foreach-object` aka `foreach` *cmdlet* alias). That was my point. Devs should use full cmdlet names ergo the only time you should see `foreach` in a script is only when using the built-in `foreach` keyword e.g. `foreach ($foo in $collection) { ... }`.

Reply:
Completely agree that folks should use full cmdlet names and avoid alias. foreach however is not an alias it is a language keyword and we should prefer that over the ForEach-Object cmdlet for most of our work because we should try and void pipelines.

Reply:
I was using `foreach` as an example only. Disregard it if it's confusing. Could also be `select` or any number of aliases returned by `get-alias`. The point was: don't use aliases in scripts. That needs to be called out. This has been a long-standing best practice for PowerShell.

Reply:
It is called out in the Style section lower down.